### ISSUE TITLE
Issues related to API (post- release of API v3)

### ISSUE THREAD

**Issue by brachbach** [oughtinc/mosaic#572]
- set whether judge-to-judge questions are allowed via the API (or default to not allowing them)
- change expert assignments on existing trees via the API
- remove trees from an experiment via the API
- set experiment as active when creating it via the API (or experiments default to being active when created)
- set experiment metadata via the API
- allow pointers to be passed down from one workspace created via the API to another (not super important, probably not worth the dev effort)
- return the IDs of the created trees (not super important)

**Comment by brachbach** [comment]
@zjmiller @andrewschreiber would any of these be easier than others?

These are the most important to me:
- set whether judge-to-judge questions are allowed via the API (or default to not allowing them)
- change expert assignments on existing trees via the API

either of those relatively easy by any chance?

**Comment by andrewschreiber** [comment]
Hmm I'd defer to @zjmiller on this one. Seems like the trick from changing expert assignments will be writing the correct GraphQL queries/mutation:
1) Provide the new expert emails as an array, and maybe root workspace id.
2) Query the existing trees of the root workspace
3) Use that Query output to generate a Mutation target for the expert assignments
4) Run the Mutation

Would need to look at how these are implemented to scope how difficult this would be particularly. I expect the debugger setup from #594 will make building these fancier API operations easier. Let me know if you think this investigation is higher priority then the issues I'm working on.

**Comment by zjmiller** [comment]
Setting judge-to-judge and changing expert assignments will both be relatively simple.

In fact, everything in the list is fairly straightforward except passing pointers around.

**Comment by brachbach** [comment]
@zjmiller could you give a time estimate for each bullet in the list, or whatever version of something like that that makes the most sense?

**Comment by zjmiller** [comment]
- [implemented in #650] set whether judge-to-judge questions are allowed via the API (or default to not allowing them)
- [2 hrs] change expert assignments on existing trees via the API
- [2 hrs] remove trees from an experiment via the API
- [1 hr] set experiment as active when creating it via the API (or experiments default to being active when created)
- [1 hr] set experiment metadata via the API
- [hard] allow pointers to be passed down from one workspace created via the API to another (not super important, probably not worth the dev effort)
- [1 hr] return the IDs of the created trees (not super important)

**Comment by brachbach** [comment]
Ok awesome!

let's do these ones:
- [2 hrs] change expert assignments on existing trees via the API
- [1 hr] return the IDs of the created trees (not super important)

And then if any of these end up being easy to tack on, then also these:
- [2 hrs] remove trees from an experiment via the API
- [1 hr] set experiment as active when creating it via the API (or experiments default to being active when created)
- [1 hr] set experiment metadata via the API

I'll change the value and dev time accordingly :)

**Comment by andrewschreiber** [comment]
Wanted to leave an update here. I've been thinking through `change expert assignments on existing trees via the API`, exploring the code, have spent more than the forecasted 2 hours on it. I think quickly writing out what I've found while trying to understand the codebase here would help get this done.

- The goal here is to be able to define a CSV file where the API user would fill out fields like `treeID, honest1Email, honest2Email, maliciousEmail`, perhaps multiple, and then a `updateExpertAssignments` command would cause the necessary changes on the backend.
- At first, my hope was that there was some `Tree` table that had an `honest1`, `honest2`, and `malicious` column. Quick update to the users (by finding the respective `User` via their email) and we're done. Turns out, not exactly.
- We do have a `Tree` object, and it instead has a many-many relationship with `UserTreeOracleRelation`. The `UserTreeOracleRelation` object has a `user` relation and a `isMalicious` bool. 
- Okay so we can get the assigned malicious expert for a given tree this way. Great. We can get the honest experts as well, but it's not clear which is honest1 and which is honest2. 
- So we need a way of differentiating between honest1 and honest2. Are timestamps too hacky, even if maybe honest1 is always made before honest2? More thought/examination needed here... 
- What does honest2 correspond to anyway? It's a role in `assignParticipants` in mosaic-api-scripts but never appears in the main mosaic repo.
- Skipping ahead, let's say we've figured out the users. We now want to replace the proper fields on the tables. So what do we update? Well, we don't know if all roles have been assigned (and the corresponding UserTreeOracleRelations have been created), so it's probably easier to delete the old UserTreeOracleRelations for the tree and create a new set, one for each newly assigned expert.
- That seems good. But wait, what is this? There is an `Assignments` table as well. It doesn't have a `treeId` attached, but it does have a `workspaceId`, a `userId`, and an `isOracle` bool. Seems pretty relevant and indeed, the `assignParticipants` command creates a large array of these. Gotta chase down what they do, probably replace them as well.
- What kind of creation cascade happens on the backend? Maybe if I assignParticipants, then we get the UserTreeOracleRelation for free? But it has to be on the right trees, which already exist and maybe already are partially responded to, so they shouldn't be deleted and recreated.
 
Stopping here as I think I've got some more investigation to do with respect to some of the points above. @zjmiller if I'm building a bridge to cross a puddle here, and there is some direct approach you had in mind, please let me know.

**Comment by zjmiller** [comment]
Whether a user is an expert for a tree depends solely on whether there exists a `UserTreeOracleRelation`. If such a relation exists but the `isMalicious` field of the relation is false, then the user is an honest expert for the tree. If such a relation exists but the `isMalicious` field is true, then the user is a malicious expert for the tree. If no such relation exists then the user is a judge for the tree.

For a while in Ought's history, "oracle" just meant "honest oracle" (and, in case you weren't aware, we used the term "oracle" for a while before we decided to use "expert" instead). So when malicious oracles were introduced, I wasn't thinking of honest and malicious oracles are two types of oracles that were conceptually on a par. I was thinking of oracles as by default honest oracles, and malicious oracles as a deviation from this default. Now, though, honest experts and malicious experts are conceptually on a par; it's weird to think of honest experts as more fundamental. So if I had to design things now, I'd have an `isHonest` boolean on the relation, instead of inferring an expert is honest from `isMalicious` being false. (So potential future refactor here).

The `isOracle` boolean on `Assignments` can be ignored. It used to be important to keep track of whether an assignment was an oracle assignment, for scheduling purposes. But things have changed a lot since then, and we aren't going back to how it was, so we could remove this field altogether as a future refactor.

So things are thankfully much simpler than you're thinking. To change expert assignments is solely a matter of creating/destroying `UserTreeOracleRelation` rows, and/or toggling the `isMalicious` field on these relations.

**Comment by zjmiller** [comment]
Also, I want to clarify the role of the scripts in the `mosaic-api-scripts` repo. These scripts are used to interact with the API endpoint of the Mosaic app. (This endpoint is the `bulkCreateExperiments` mutation of the GraphQL schema.)

The scripts in `mosaic-api-scripts` don't really have anything to do with Mosaic backend code. The scripts in `mosaic-api-scripts` are all just ways to more efficiently create requests to be sent to the Mosaic API endpoint.

So, looking at your second-to-last bullet point, the `assignParticipants` script is just something on Ben's computer or Mati's computer that helps them interact with the Mosaic API endpoint. I haven't even looked at the script personally, as the ecosystem that has grown up to interact with the Mosaic API endpoint has become quite large. And I don't really need to look at it, because everyone should feel free to create whatever personal scripts/tools help them engage with the endpoint more efficiently.  

In Mosaic proper, the concept of an assignment is consistently used to mean the assignment of a user to a workspace during an experiment. But if Ben or Mati want to make their own script, to be run on their own computer, and use that concept to mean something else, that's fine by me. So don't take anything in `mosaic-api-scripts` too seriously as a means to better understanding the internals of Mosaic proper.

Relatedly, if a script from `mosaic-api-scripts` creates a bunch of "assignment" objects, these are just entities on Ben's computer or Mati's computer, and these objects are assisting them to create a GraphQL request to send to the Mosaic API endpoint. There's no sense in which these entities exist in the actual Mosaic backend. Indeed, the Mosaic API endpoint never modifies the Assignments table in any way.

So, returning to your second-to-last bullet point, the `assignParticipants` script doesn't actually create any rows in the `Assignments` table. This comment should also help illustrate why the `assignParticipants` script has honest1 and honest2 references but these don't appear in the Mosaic backend. The scripts and the Mosaic app are just two separate things.

**Comment by andrewschreiber** [comment]
Thanks for the detailed response! 

I particularly appreciate the history around the term `oracle`, how it's evolved as Mosaic has changed, and the degree to which the current state of our object models has substantial influence from the older UX. I'm also taking away that the API scripts are more 'bolted on' than designed in-tandem with Mosaic, and will lower my expectation that the terminology shares semantics across repos.

**Comment by Mati-Roy** [comment]
meta: I haven't caught out with this thread, but I just posted the following on Slack, and realized I should probably post here instead.

tl;dr / bottom line: it seems we now need the API to support updating tree assignment

Anna just finished L1. I was going to assign Anna to 10 trees as HE and 10 trees as ME in L2 to get them started right away, but I just realized that the proper way to do it would be to assign them to one-third of all trees for each role (or something like that) with alternating priority (ex.: HE for the tree with priority 1, ME for the P2, J for the P3, etc.) in order to avoid them having only one role for long periods of time. Just adding more trees to assign Anna would result in Anna being judges on a bunch of higher priority trees. We could add a series of tree with incrementing schedulingPriority starting with 1, but a) everytime we'd do that, we would push back the date at which the current trees would be completed, and b) this wouldn't allow us to (cleanly) add participants as they graduate from L1 (we'd likely want need to wait to add a bunch at the same time in that situation). And with 248 trees (soon 1000) and more L1 participants approaching graduation, doing it manually doesn't seem like a viable / optimal option at all.
